### PR TITLE
ExpansionTile: support tween colors for header and icon

### PR DESCRIPTION
**Pull request by arthurdenner**
## Description
Resubmission of #37262 but with tests and updated documentation.
I decided to open a new PR as it seems the original authors' fork is gone.
Still, I'd like to thank him for the suggested changes.

## Related Issues
N/A

## Tests

I added the following tests:

- Specified tween colors are applied properly.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner. (I'm trying to)

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by arthurdenner** [comment]
cc @Piinks.

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 1]
```diff
@@ -80,6 +84,22 @@ class ExpansionTile extends StatefulWidget {
   /// Specifies if the list tile is initially expanded (true) or collapsed (false, the default).
   final bool initiallyExpanded;
 
+  /// The color of the header when the tile is collapsed.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/expansion_tile.dart]: It might actually be better to simply pass in a [ColorTween](https://master-api.flutter.dev/flutter/animation/ColorTween-class.html). For example:
```dart
/// Docs
final ColorTween headerColorTween;

/// Docs
final ColorTween iconColorTween;
```

This makes the API simpler since there are only two parameters instead of four.

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: This might require moving some refactoring of _ExpansionTileState._headerColorTween and _ExpansionTileState._iconColorTween, since they're currently final fields.

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 2]
```diff
@@ -229,4 +229,69 @@ void main() {
...
+    expect(iconColor(expandedIconKey), expandedIconColor);
+    expect(iconColor(collapsedIconKey), collapsedIconColor);
+
+    // Tap both tiles to change their state: collapse and expand respectively
+    await tester.tap(find.text('Expanded'));
+    await tester.tap(find.text('Collapsed'));
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
+
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: We should probably test at least one color value mid-transition as well to ensure that the the color is properly lerp-ing from start to finish. A quick example I could find is [here](https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/theme_test.dart#L289). Check and see how long the transition takes and jump halfway through it, making sure the color is the expected value.

File: packages/flutter/lib/src/material/expansion_tile.dart (hunk 1 above)
Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: @shihaohong, pass a `ColorTween` doesn't seem to be the default implementation on other widgets, e.g, `BottomNavigationBar` takes `unselectedItemColor` and `selectedItemColor`.

What I noticed is that they use more semantic names, which we could apply here. What do you think? I've thought about the following names (same I used on the tests):

- expandedTitleColor
- expandedIconColor
- collapsedTitleColor
- collapsedIconColor

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: The internals of the widget still end up using a `ColorTween`, so I feel like it does make the API cleaner without forcing anything new onto the user. We could add some API documentation that they could use the `ColorTween` to simply toggle a color if they wanted to as well. 

A key difference would be that we would be replacing the entire `ColorTween` instead of updating values on a single existing `ColorTween`. There might be implications to that that I do not understand. @HansMuller do you know if this might cause any problems or observable differences in behavior?

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Found a section of the [Tween API docs](https://master-api.flutter.dev/flutter/animation/Tween-class.html) on mutable Tweens. Making this type of change could have a performance implication

Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: I see. Should we wait for @HansMuller's input or can I proceed with the renaming properties approach?

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Do you mind testing to see if there are any observable differences between the two? Just building both out on a device and seeing if performance is impacted, does changing the tween’s colors do anything different, etc

Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: I'll try to do that in the following days.

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Thank you! Let me know if you need any help or suggestions

Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: I've updated the PR with your suggested changes and test them using this stateful version of the [expansion_tile_sample.dart example](https://gist.github.com/arthurdenner/1780d3594a06eb06b94474d45c23301a).

Performance doesn't seem to be affected, but the icon is not updating when calling `setState`. If I copy the logic from `_ExpansionTileState.didChangeDependencies` to `_ExpansionTileState.didUpdateWidget`, they do, but I'm not sure about this change, so I didn't commit it.

If I was clear on the explanation, what are your thoughts on this? I can commit these changes so that you can test too if you prefer and remove them before the merge.

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: This makes sense. `didChangeDependencies` is called immediately after `initState`, and in the original implementation, a private `ColorTween` is created in `initState` and its begin and end values were set to the passed in `ColorTween`s begin and end values.

On the other hand, `didUpdateWidget` is called after every `build`, which is necessary here if you ever update ColorTween being used using `setState`.

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: It's pretty dense, but the [State](https://api.flutter.dev/flutter/widgets/State-class.html) documentation pretty much outlines the lifecycle for State objects.

Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Nice, thanks! I'll read it. Should I commit and push these changes then?

Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Go ahead and commit it so I (and whoever else that reads this PR) can take a look and see if it works out :)

Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Done. I've updated the `expansion_tile_sample` example too for an easier demo.

File: examples/catalog/lib/expansion_tile_sample.dart [hunk 3]
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:math';
```
Comment (shihaohong) [inline examples/catalog/lib/expansion_tile_sample.dart]: I would revert the sample changes and make it a separate PR, since I'm not sure if this is what the maintainers of the catalog would want

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 4]
```diff
@@ -80,6 +82,16 @@ class ExpansionTile extends StatefulWidget {
   /// Specifies if the list tile is initially expanded (true) or collapsed (false, the default).
   final bool initiallyExpanded;
 
+  /// The tween to control the color of the header.
+  /// begin defaults to theme.textTheme.subtitle1.color
+  /// end defaults to theme.accentColor
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/expansion_tile.dart]: Just shifting the wording a bit here:
```
  /// The tween that controls the color of the header.
  ///
  /// This tween's `begin` property defaults to [ThemeData.textTheme.subtitle1.color],
  /// whereas the tween's `end` property defaults to [ThemeData.accentColor].
```

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 5]
```diff
@@ -80,6 +82,16 @@ class ExpansionTile extends StatefulWidget {
...
+  final ColorTween headerColorTween;
+
+  /// The tween to control the color of the header.
+  /// begin defaults to theme.unselectedWidgetColor
+  /// end defaults to theme.accentColor
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/expansion_tile.dart]: /// The tween that controls the icon's color.
  ///
  /// This tween's `begin` property defaults to [ThemeData.unselectedWidgetColor],
  /// whereas the tween's `end` property defaults to [ThemeData.accentColor].

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 6]
```diff
@@ -187,20 +201,30 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
...
+        ..end = theme.dividerColor;
+      _headerColorTween
+        ..begin = widget.headerColorTween?.begin ?? theme.textTheme.subtitle1.color
+        ..end = widget.headerColorTween?.end ?? theme.accentColor;
+      _iconColorTween
+        ..begin = widget.iconColorTween?.begin ?? theme.unselectedWidgetColor
+        ..end = widget.iconColorTween?.end ?? theme.accentColor;
+      _backgroundColorTween
+        ..end = widget.backgroundColor;
+    }
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/expansion_tile.dart]: Indentation here is off by two spaces

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 7]
```diff
@@ -229,4 +229,145 @@ void main() {
...
+      end: expandedIconColor,
+    );
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Material(
+          child: SingleChildScrollView(
+            child: Column(
+              children: <Widget>[
+                const ListTile(title: Text('Top')),
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Is this `ListTile` necessary?

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 8]
```diff
@@ -229,4 +229,145 @@ void main() {
...
+    await tester.tap(find.text('Collapsed'));
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
+
+    expect(textColor(expandedTitleKey), collapsedTitleColor);
+    expect(textColor(collapsedTitleKey), expandedTitleColor);
+    expect(iconColor(expandedIconKey), collapsedIconColor);
+    expect(iconColor(collapsedIconKey), expandedIconColor);
+  }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }));
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Why test on only iOS and macOS? We could just remove the variant property right?

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 9]
```diff
@@ -229,4 +229,145 @@ void main() {
...
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Same comment about testing on only iOS and macOS variants

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 10]
```diff
@@ -229,4 +229,145 @@ void main() {
 
     expect(find.text('Subtitle'), findsOneWidget);
   });
+
+  testWidgets('ExpansionTile custom tween colors', (WidgetTester tester) async {
+    final Key expandedTitleKey = UniqueKey();
+    final Key collapsedTitleKey = UniqueKey();
+    final Key expandedIconKey = UniqueKey();
+    final Key collapsedIconKey = UniqueKey();
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: I think the naming here is confusing because the tiles end up expanding and collapsing over time. Could we just use a single `ExpansionTile` and test the custom icon and title colors on that? In other words:
```dart
                ExpansionTile(
                  initiallyExpanded: true,
                  headerColorTween: headerColorTween,
                  iconColorTween: iconColorTween,
                  title: TestText('Expanded', key: titleKey),
                  children: const <Widget>[ListTile(title: Text('0'))],
                  trailing: TestIcon(key: iconKey),
                ),
```
just this single ExpansionTile and testing the custom colors when the tile is open and closed.

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 11]
```diff
@@ -229,4 +229,145 @@ void main() {
...
+    await tester.pump(const Duration(seconds: 1));
+
+    expect(textColor(expandedTitleKey), collapsedTitleColor);
+    expect(textColor(collapsedTitleKey), expandedTitleColor);
+    expect(iconColor(expandedIconKey), collapsedIconColor);
+    expect(iconColor(collapsedIconKey), expandedIconColor);
+  }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }));
+
+  testWidgets('ExpansionTile mixed custom and default tween colors', (WidgetTester tester) async {
+    final Key expandedTitleKey = UniqueKey();
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Same comment about using a single `ExpansionTile`. Maybe test setting just `begin` for one of the tweens and just `end` for the other tween.

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 12]
```diff
@@ -229,4 +229,145 @@ void main() {
...
+    const Color collapsedTitleColor = Colors.red;
+    const Color expandedIconColor = Colors.green;
+    const Color collapsedIconColor = Colors.orange;
+    final ColorTween headerColorTween = ColorTween(
+      begin: collapsedTitleColor,
+    );
+    final ColorTween iconColorTween = ColorTween(
+      end: expandedIconColor,
+    );
+    final ThemeData theme = ThemeData(
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Maybe add a comment here that `The theme's `accentColor` and `unselectedWidgetColor` are used when ...`

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 13]
```diff
@@ -229,4 +229,145 @@ void main() {
...
+    );
+
+    await tester.pumpWidget(
+      MaterialApp(
+        theme: theme,
+        home: Material(
+          child: SingleChildScrollView(
+            child: Column(
+              children: <Widget>[
+                const ListTile(title: Text('Top')),
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Same comment about having the ListTile present in the test

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 14]
```diff
@@ -229,4 +229,145 @@ void main() {
...
+            ),
+          ),
+        ),
+      ),
+    );
+
+    Color iconColor(Key key) => tester.state<TestIconState>(find.byKey(key)).iconTheme.color;
+    Color textColor(Key key) => tester.state<TestTextState>(find.byKey(key)).textStyle.color;
+
+    expect(textColor(expandedTitleKey), expandedTitleColor);
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Maybe compare with the expected color like:
```dart
    expect(textColor(expandedTitleKey), Colors.blue);
```

I find this easier to follow since my immediate assumption was that this test was simply checking to see if its color was equal to itself.

File: examples/catalog/lib/expansion_tile_sample.dart (hunk 3 above)
Comment (arthurdenner) [reply examples/catalog/lib/expansion_tile_sample.dart]: I wasn't planning to leave it, it was just for testing, like I mentioned here: `I can commit these changes so that you can test too if you prefer and remove them before the merge.`

File: packages/flutter/lib/src/material/expansion_tile.dart (hunk 6 above)
Comment (arthurdenner) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Any tips on how to format the code properly and maybe automatically? I use VSCode and have Prettier installed, but whenever I save a file, it formats more than I want. So I turn it off and these kind of things pass :/

File: examples/catalog/lib/expansion_tile_sample.dart (hunk 3 above)
Comment (arthurdenner) [reply examples/catalog/lib/expansion_tile_sample.dart]: I've removed the changes to avoid confusion as they are available [here](https://gist.github.com/arthurdenner/1780d3594a06eb06b94474d45c23301a).

File: packages/flutter/test/material/expansion_tile_test.dart (hunk 2 above)
Comment (arthurdenner) [reply packages/flutter/test/material/expansion_tile_test.dart]: Do you still think this is valid or are the current tests enough?

File: examples/catalog/lib/expansion_tile_sample.dart (hunk 3 above)
Comment (shihaohong) [reply examples/catalog/lib/expansion_tile_sample.dart]: Sorry about that, when I read your initial comment I thought you were referencing your implementation of the header and icon tweens and not the sample!

File: packages/flutter/lib/src/material/expansion_tile.dart (hunk 6 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/expansion_tile.dart]: I don't think there is a good way unless you implement your own formatting rules. I currently also just format code without any help from tools like Prettier

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 15]
```diff
@@ -229,4 +229,93 @@ void main() {
...
+    Color textColor(Key key) => tester.state<TestTextState>(find.byKey(key)).textStyle.color;
+
+    expect(iconColor(iconKey), Colors.green);
+    expect(textColor(titleKey), Colors.blue);
+
+    // Tap the tile to change its state to collapsed
+    await tester.tap(find.text('Expanded'));
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
```
Comment (shihaohong) [inline packages/flutter/test/material/expansion_tile_test.dart]: Why are there two pump calls with a duration of 1 second for each here?

File: packages/flutter/test/material/expansion_tile_test.dart (hunk 2 above)
Comment (shihaohong) [reply packages/flutter/test/material/expansion_tile_test.dart]: Yeah, I still think it's important since there isn't any part of the test that checks for the animating colors. It should be pretty straightforward to implement, but let me know if you run into any issues!

File: packages/flutter/test/material/expansion_tile_test.dart (hunk 15 above)
Comment (arthurdenner) [reply packages/flutter/test/material/expansion_tile_test.dart]: Honestly, I have no idea :)

This test case is based on the `ListTileTheme` from the same file. I tried to keep the same structure. I always try to do so and change if requested 😅

I tried to remove one of these during the last refactor but wasn't able to do so, tests were failing. But thanks for bringing it up here, I checked other tests and replaced these `.pump` calls with `.pumpAndSettle`. Hope this is a good practice - I'm still quite new to testing in Flutter/Dart, have just started with unit tests.

Comment (shihaohong) [reply packages/flutter/test/material/expansion_tile_test.dart]: I did some additional digging, and it seems like there is an issue with trying to update the header color. This works:
```dart
    final Key iconKey = UniqueKey();
    final Key titleKey = UniqueKey();
    final ColorTween headerColorTween = ColorTween(
      begin: Colors.red,
    );
    final ColorTween iconColorTween = ColorTween(
      end: Colors.green,
    );
    /// ThemeData.accentColor is used as the default value for headerColorTween.end
    /// ThemeData.unselectedWidgetColor is used as the default value for iconColorTween.begin
    final ThemeData theme = ThemeData(
      accentColor: Colors.blue,
      unselectedWidgetColor: Colors.orange,
    );

    await tester.pumpWidget(
      MaterialApp(
        theme: theme,
        home: Material(
          child: ExpansionTile(
            initiallyExpanded: true,
            headerColorTween: headerColorTween,
            iconColorTween: iconColorTween,
            title: TestText('Expanded', key: titleKey),
            trailing: TestIcon(key: iconKey),
          ),
        ),
      ),
    );

    Color iconColor(Key key) => tester.state<TestIconState>(find.byKey(key)).iconTheme.color;
    expect(iconColor(iconKey), Colors.green);

    // Tap the tile to change its state to collapsed
    await tester.tap(find.text('Expanded'));
    await tester.pump();
    await tester.pump(const Duration(milliseconds: 100));

    // The icon's color transitions with an ease in curve.
    expect(iconColor(iconKey), const Color(0xffc69f19));
    await tester.pump(const Duration(milliseconds: 100));

    // Fully transitions to end color.
    expect(iconColor(iconKey), Colors.orange);
```

However, tweaking the numbers on `textColor`, it seems that it takes 400ms or longer for the color changes to reach the endpoint and it has intermediary colors that didn't make sense to me. My hunch is that there is some strange interaction between this tween and some underlying animation mechanism.

I think we should reconsider this change until we understand why this is happening. Does that sound okay with you?

Comment (shihaohong) [reply packages/flutter/test/material/expansion_tile_test.dart]: I thought about adding just the icon tween, but am a little concerned about that idea since we don't understand why that one works but the header tween does not.

Comment (arthurdenner) [reply packages/flutter/test/material/expansion_tile_test.dart]: I'd like to have the icon as customizable but I completely understand your points and I have a workaround in place. Feel free to close the PR from my side. 👍

**Comment by shihaohong** [comment]
Closing this PR because it is not known why the `headerColor` tween is not applied in the duration expected (200ms at the time of this comment), so until we can figure out why it's taking longer and transitioning between colors that aren't expected, we've decided to not move forward with icon and header tween customization.

See the above threads for more details!
